state.go

```diff
@@ -272,10 +272,21 @@ func (m *Memberlist) probeNode(node *nodeState) {
 	// soon as possible.
 	deadline := sent.Add(probeInterval)
 	addr := node.Address()
+
+	// Arrange for our self-awareness to get updated.
+	var awarenessDelta int
+	defer func() {
+		m.awareness.ApplyDelta(awarenessDelta)
+	}()
 	if node.State == stateAlive {
 		if err := m.encodeAndSendMsg(addr, pingMsg, &ping); err != nil {
```

Comment:
An alternative to the `ProtocolType() ProtocolType` interface change would be to have the `WriteTo()` method return a special error if the send has hard failed (in the case of TCP), maybe something like:
```
type TransportError struct {
  Err      error
  HardFailure bool // bikeshed this term; this used just for example here
}

// for go 1.13 errors.Unwrap() compatibility
func (e *TransportError) Unwrap() error { return e.Err }

// so the underlying error is what is printed if pretty printed
func (e *TransportError) Error() string { return e.Err.Error() }
```

Then (pre go 1.13) you can check:
```
err := m.encodeAndSendMsg(...)
if err != nil {
  // ... your current logging logic (the Error() impl above will imply the messages are printed identically) ...
  if terr, ok := err.(*TransportError); ok && terr.HardFailure {
    // ... your current TCP logic
  } else {
   // ... your current UDP logic
  }
}
```

Reply:
If you do this, I would say your `goto` target should probably be named something like `HANDLE_HARDFAIL` or something appropriately named.

I sorta wish there were a straightforward way to avoid the goto.

Reply:
I renamed the label to `HANDLE_REMOTE_FAILURE` and I am using `net.OpError` to decided what to do. That way the transport doesn't have to change.